openmc/tallies.py

```diff
@@ -0,0 +1,3204 @@
+from collections.abc import Iterable, MutableSequence
+import copy
+import re
+from functools import partial, reduce
+from itertools import product
+from numbers import Integral, Real
+import operator
+from pathlib import Path
+import warnings
+from xml.etree import ElementTree as ET
+
+import numpy as np
+import pandas as pd
+import scipy.sparse as sps
+import h5py
+
+import openmc
+import openmc.checkvalue as cv
+from openmc._xml import clean_indentation
+from .mixin import IDManagerMixin
+
+
+# The tally arithmetic product types. The tensor product performs the full
+# cross product of the data in two tallies with respect to a specified axis
+# (filters, nuclides, or scores). The entrywise product performs the arithmetic
+# operation entrywise across the entries in two tallies with respect to a
+# specified axis.
+_PRODUCT_TYPES = ['tensor', 'entrywise']
+
+# The following indicate acceptable types when setting Tally.scores,
+# Tally.nuclides, and Tally.filters
+_SCORE_CLASSES = (str, openmc.CrossScore, openmc.AggregateScore)
+_NUCLIDE_CLASSES = (str, openmc.CrossNuclide, openmc.AggregateNuclide)
+_FILTER_CLASSES = (openmc.Filter, openmc.CrossFilter, openmc.AggregateFilter)
+
+# Valid types of estimators
+ESTIMATOR_TYPES = ['tracklength', 'collision', 'analog']
+
+
+class Tally(IDManagerMixin):
+    """A tally defined by a set of scores that are accumulated for a list of
+    nuclides given a set of filters.
+
+    Parameters
+    ----------
+    tally_id : int, optional
+        Unique identifier for the tally. If none is specified, an identifier
+        will automatically be assigned
+    name : str, optional
+        Name of the tally. If not specified, the name is the empty string.
+
+    Attributes
+    ----------
+    id : int
+        Unique identifier for the tally
+    name : str
+        Name of the tally
+    filters : list of openmc.Filter
+        List of specified filters for the tally
+    nuclides : list of openmc.Nuclide
```

Comment:
After looking through a lot of the validation merge logic, it seems like we could define a new type that each of `filters`, `nuclides`, and `scores` share that enables us to prevent duplicates but also defines the logic that can be used to merge the sets together.